### PR TITLE
fix(external-plugins)  response.getXXX() functions from log phase

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
The `kong.response.get_XXX()` functions require data from the response
object, which is no longer accessible in the "post-log" timer used to
call Log handlers in external plugins.

This patch allows those functions to work by adding the required data to
 the set saved at the start of the "real" log phase.

Fix Kong/go-pdk#61

### CODE REVIEW

File: kong/runloop/plugin_servers/init.lua [hunk 1]
```diff
@@ -15,6 +17,75 @@ local rpc_notifications = {}
...
+local exposed_api = {
+  kong = kong,
+
+  ["kong.log.serialize"] = function()
+    local saved = save_for_later[coroutine.running()]
+    return cjson.encode(saved and saved.serialize_data or kong.log.serialize())
+  end,
+
+  ["kong.nginx.get_var"] = function(v)
+    return ngx.var[v]
```
Comment (dndx) [inline kong/runloop/plugin_servers/init.lua:29]: `ngx.var` is something we can cache at file level, this speed up this potentially hot call path a little considering variable access are notoriously slow in OpenResty already.

File: kong/runloop/plugin_servers/init.lua [hunk 2]
```diff
@@ -15,6 +17,75 @@ local rpc_notifications = {}
 --- currently running plugin instances
 local running_instances = {}
 
+local exposed_api = {
+  kong = kong,
+
+  ["kong.log.serialize"] = function()
+    local saved = save_for_later[coroutine.running()]
```
Comment (dndx) [inline kong/runloop/plugin_servers/init.lua:24]: We should be able to cache `coroutine.running` as well to speed up the hot path.

**Review by dndx — APPROVED** [review]
Minor comments, otherwise LGTM.

File: kong/runloop/plugin_servers/init.lua (hunk 1 above)
Comment (javierguerragiraldez) [reply kong/runloop/plugin_servers/init.lua:29]: this is to allow an external plugin (Go, JS) to do `ngx.var[]`;  caching could defeat the purpose and the performance issue is much less than the RPC overhead.

File: kong/runloop/plugin_servers/init.lua (hunk 2 above)
Comment (javierguerragiraldez) [reply kong/runloop/plugin_servers/init.lua:24]: it's used to differentiate the request that is being handled.  the problem is that the `log` phase is handled in a timer context, after the request object itself has been released, so we have to associate those values with the timer instance.  I remember looking for some timer ID but didn't find anything usable.  I'll check again.

**Review by dndx — APPROVED** [review]

**Pull request merged**